File: lab04/trapecioPool.cpp
#include <iostream>
#include <vector>
#include <thread>
#include <mutex>
#include <condition_variable>
#include <queue>
#include <future>
#include <iomanip>
#include <sstream>
#include <chrono>
#include <limits>
#include <algorithm>
#include <type_traits>

using namespace std;

struct Funcion {
    virtual double eval(double x) const = 0;
    virtual ~Funcion() = default;
};

struct FuncionCuadratica : Funcion {
    double eval(double x) const override {
        return 2.0 * x * x + 3.0 * x + 0.5;
    }
};

class Pool {
public:
    explicit Pool(size_t n) : detener(false) {
        n = max<size_t>(1, n);
        for (size_t i = 0; i < n; ++i) {
            trabajadores.emplace_back([this]() {
                for (;;) {
                    function<void()> tarea;
                    {
                        unique_lock<mutex> lk(mtx);
                        cv.wait(lk, [this]() { return detener || !cola.empty(); });
                        if (detener && cola.empty()) return;
                        tarea = move(cola.front());
                        cola.pop();
                    }
                    tarea();
                }
            });
        }
    }

    template <class F>
    auto enviar(F&& f) -> future<invoke_result_t<F&>> {
        using R = invoke_result_t<F&>;
        packaged_task<R()> task(forward<F>(f));
        auto fut = task.get_future();
        {
            unique_lock<mutex> lk(mtx);
            if (detener) throw runtime_error("pool detenido");
            cola.emplace([t = move(task)]() mutable { t(); });
        }
        cv.notify_one();
        return fut;
    }

    ~Pool() {
        {
            unique_lock<mutex> lk(mtx);
            detener = true;
        }
        cv.notify_all();
        for (auto& w : trabajadores) w.join();
    }

private:
    vector<thread> trabajadores;
    queue<function<void()>> cola;
    mutex mtx;
    condition_variable cv;
    bool detener;
};

class Integrador {
public:
    Integrador(const Funcion& f, size_t hilos)
        : f(f), pool(hilos), hilos(hilos) {}

    double integrar(double a, double b, long long n) {
        double h = (b - a) / static_cast<double>(n);
        long long interior = max(0LL, n - 1);
        long long base = interior / static_cast<long long>(hilos);
        long long resto = interior % static_cast<long long>(hilos);

        vector<future<double>> futs;
        futs.reserve(hilos);

        long long inicio = 1;
        for (size_t w = 0; w < hilos; ++w) {
            long long tam = base + (static_cast<long long>(w) < resto ? 1 : 0);
            long long fin = (tam == 0) ? (inicio - 1) : (inicio + tam - 1);

            auto tarea = [this, a, h, inicio, fin]() -> double {
                double suma = 0.0;
                for (long long i = inicio; i <= fin; ++i) {
                    double x = a + i * h;
                    suma += f.eval(x);
                }
                return suma;
            };

            futs.emplace_back(pool.enviar(tarea));
            inicio = fin + 1;
        }

        double suma = 0.0;
        for (auto& fu : futs) suma += fu.get();

        double fa = f.eval(a);
        double fb = f.eval(b);
        return h * (0.5 * fa + suma + 0.5 * fb);
    }

private:
    const Funcion& f;
    Pool pool;
    size_t hilos;
};

static string clave(double v) {
    ostringstream oss;
    oss.setf(std::ios::fixed);
    oss << setprecision(12) << v;
    return oss.str();
}

int main(int argc, char** argv) {
    double a = (argc >= 2) ? atof(argv[1]) : 2.0;
    double b = (argc >= 3) ? atof(argv[2]) : 20.0;
    long long Nmax = (argc >= 4) ? atoll(argv[3]) : 2000000LL;
    size_t hilos = (argc >= 5) ? static_cast<size_t>(atoll(argv[4]))
                               : max(1u, thread::hardware_concurrency());

    FuncionCuadratica f;
    Integrador integ(f, hilos);

    cout.setf(std::ios::fixed);
    cout << setprecision(12);

    string ultima;
    double valor = numeric_limits<double>::quiet_NaN();
    long long repetido = -1;

    auto t0 = chrono::steady_clock::now();
    for (long long n = 1; n <= Nmax; ++n) {
        double v = integ.integrar(a, b, n);
        string k = clave(v);
        if (n == 1) {
            ultima = k;
            valor = v;
        } else {
            if (k == ultima) {
                repetido = n;
                valor = v;
                break;
            }
            ultima = k;
            valor = v;
        }
        if (n <= 10 || n % 100000 == 0) {
            cout << "n=" << n << " -> " << v << "\n";
        }
    }
    auto t1 = chrono::steady_clock::now();
    double s = chrono::duration<double>(t1 - t0).count();

    if (repetido > 0) {
        cout << "Repetido en n=" << repetido << " -> " << valor << "\n";
    } else {
        cout << "Finalizado -> " << valor << "\n";
    }
    cout << "Tiempo: " << s << " s\n";
    return 0;
}

File: lab04/TrapecioPool.java
import java.util.*;
import java.util.concurrent.*;
import java.math.BigDecimal;
import java.math.RoundingMode;

interface Funcion {
    double evaluar(double x);
}

final class FuncionCuadratica implements Funcion {
    public double evaluar(double x) { return 2.0*x*x + 3.0*x + 0.5; }
}

final class TareaTrapecio implements Callable<Double> {
    private final Funcion f;
    private final double a;
    private final double h;
    private final long inicio, fin;

    TareaTrapecio(Funcion f, double a, double h, long inicio, long fin) {
        this.f = f; this.a = a; this.h = h; this.inicio = inicio; this.fin = fin;
    }

    @Override public Double call() {
        double suma = 0.0;
        for (long i = inicio; i <= fin; i++) {
            double x = a + i * h;
            suma += f.evaluar(x);
        }
        return suma;
    }
}

final class IntegradorTrapecio {
    private final Funcion f;
    private final ExecutorService pool;
    private final int hilos;

    IntegradorTrapecio(Funcion f, int hilos) {
        this.f = f;
        this.hilos = Math.max(1, hilos);
        this.pool = Executors.newFixedThreadPool(this.hilos);
    }

    public double integrar(double a, double b, long n) throws Exception {
        final double h = (b - a) / n;
        List<Future<Double>> futuros = new ArrayList<>(hilos);
        long interior = Math.max(0, n - 1);
        long base = interior / hilos;
        long resto = interior % hilos;
        long inicio = 1;
        for (int w = 0; w < hilos; w++) {
            long tam = base + (w < resto ? 1 : 0);
            long fin = (tam == 0) ? (inicio - 1) : (inicio + tam - 1);
            futuros.add(pool.submit(new TareaTrapecio(f, a, h, inicio, fin)));
            inicio = fin + 1;
        }
        double suma = 0.0;
        for (Future<Double> fu : futuros) suma += fu.get();
        double fa = f.evaluar(a);
        double fb = f.evaluar(b);
        return h * (0.5 * fa + suma + 0.5 * fb);
    }

    public void cerrar() { pool.shutdown(); }
}

public class TrapecioPool {
    private static String clave(double v) {
        return new BigDecimal(v).setScale(12, RoundingMode.HALF_UP).toPlainString();
    }

    public static void main(String[] args) throws Exception {
        double a = (args.length >= 1) ? Double.parseDouble(args[0]) : 2.0;
        double b = (args.length >= 2) ? Double.parseDouble(args[1]) : 20.0;
        long Nmax = (args.length >= 3) ? Long.parseLong(args[2]) : 2_000_000L;
        int hilos = (args.length >= 4) ? Integer.parseInt(args[3]) : Math.max(1, Runtime.getRuntime().availableProcessors());

        Funcion f = new FuncionCuadratica();
        IntegradorTrapecio integ = new IntegradorTrapecio(f, hilos);

        String ultimaClave = null;
        double ultimoValor = Double.NaN;
        long repetido = -1;
        long n;
        long t0 = System.nanoTime();
        for (n = 1; n <= Nmax; n++) {
            double val = integ.integrar(a, b, n);
            String k = clave(val);
            if (k.equals(ultimaClave)) { repetido = n; ultimoValor = val; break; }
            ultimaClave = k;
            ultimoValor = val;
            if (n <= 10 || n % 100000 == 0) {
                System.out.printf(Locale.US, "n=%d -> %.12f%n", n, val);
            }
        }
        long t1 = System.nanoTime();

        if (repetido > 0) {
            System.out.printf(Locale.US, "Repetido en n=%d -> %.12f%n", repetido, ultimoValor);
        } else {
            System.out.printf(Locale.US, "Finalizado en n=%d -> %.12f%n", n - 1, ultimoValor);
        }
        System.out.printf(Locale.US, "Tiempo: %.3f s%n", (t1 - t0)/1e9);
        integ.cerrar();
    }
}
